test/unit/resource/tests/test-resource-BasicAsyncSemantics.hpp

```diff
@@ -0,0 +1,69 @@
+//~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~//
+// Copyright (c) 2016-22, Lawrence Livermore National Security, LLC
+// and RAJA project contributors. See the RAJA/LICENSE file for details.
+//
+// SPDX-License-Identifier: (BSD-3-Clause)
+//~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~//
+
+#ifndef __TEST_RESOURCE_DEPENDS_HPP__
+#define __TEST_RESOURCE_DEPENDS_HPP__
+
+#include "RAJA_test-base.hpp"
+
+template <typename WORKING_RES, typename EXEC_POLICY>
+void ResourceBasicAsyncSemanticsTestImpl()
+{
+  constexpr std::size_t ARRAY_SIZE{10000000};
+  using namespace RAJA;
+
+  WORKING_RES dev;
+  resources::Host host;
+
+  int* d_array = resources::Resource{dev}.allocate<int>(ARRAY_SIZE);
+  int* h_array  = host.allocate<int>(ARRAY_SIZE);
+
+  forall<policy::sequential::seq_exec>(host, RangeSegment(0,ARRAY_SIZE),
+    [=] RAJA_HOST_DEVICE (int i) {
+      h_array[i] = i;
+    }
+  );
+
+  dev.memcpy(d_array, h_array, sizeof(int) * ARRAY_SIZE);
+
+  forall<EXEC_POLICY>(dev, RangeSegment(0,ARRAY_SIZE),
+    [=] RAJA_HOST_DEVICE (int i) {
+      d_array[i] = i + 2;
+    }
+  );
+
+  dev.memcpy(h_array, d_array, sizeof(int) * ARRAY_SIZE);
```

Comment:
Probably want a wait between this and the forall right?

Reply:
Done.